### PR TITLE
Added controller for PVC deletion

### PR DESCRIPTION
**Pull request by Danil-Grigorev**
/kind feature

**What this PR does / why we need it**:

This PR addresses an attempt to split PVC handling between `external-provisioner` and `sig-storage-lib-external-provisioner`, also introducing a way to handle finalizers, related to in-progress PVC cloning.

PVC referenced in a data source is now updated with finalizer `provisioner.storage.kubernetes.io/cloning-protectionr` at the moment when a `ProvisionExt` method is called, used to dynamically provision a PV to which the content will be cloned at.

When cloning is finished - all PVC referencing the one as
a data source goes from `Pending` to `Bound` state, which allows the finalizer to be removed before deletion.
 
**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes-csi/external-provisioner/issues/414

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
finalizers:
- provisioner.storage.kubernetes.io/cloning-protection

is now set on PVCs which are the source ones for the cloning,
preventing their removal before the cloning finishes.
```

### CODE REVIEW

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -531,6 +582,13 @@ func (p *csiProvisioner) ProvisionExt(options controller.ProvisionOptions) (*v1.
 		req.VolumeContentSource = volumeContentSource
 	}
 
+	if options.PVC.Spec.DataSource != nil && rc.clone {
+		err = p.setCloneFinalizer(options.PVC.Spec.DataSource.Name, options.PVC.Namespace)
```
Comment (jsafrane) [inline pkg/controller/controller.go]: Is there reason you pass name/namespace instead of full PVC? `setCloneFinalizer` does not need to query API server at all.

File: pkg/controller/controller.go [hunk 2]
```diff
@@ -997,6 +1070,150 @@ func (p *csiProvisioner) Delete(volume *v1.PersistentVolume) error {
...
+		var key string
+		var ok bool
+		if key, ok = obj.(string); !ok {
+			p.claimQueue.Forget(obj)
+			return fmt.Errorf("expected string in workqueue but got %#v", obj)
+		}
+
+		if err := p.syncClaimHandler(key); err != nil {
+			klog.Warningf("Retrying syncing claim %q after %v failures", key, p.claimQueue.NumRequeues(obj))
+			p.claimQueue.AddRateLimited(obj)
```
Comment (jsafrane) [inline pkg/controller/controller.go]: You should not call `Forget` below if you want the obj in the queue.

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -997,6 +1070,150 @@ func (p *csiProvisioner) Delete(volume *v1.PersistentVolume) error {
...
+	claim, ok := obj.(*v1.PersistentVolumeClaim)
+	if !ok {
+		return fmt.Errorf("expected claim but got %+v", obj)
+	}
+
+	if !checkFinalizer(claim, pvcCloneFinalizer) {
+		return nil
+	}
+
+	pvcList, err := p.client.CoreV1().PersistentVolumeClaims("").List(metav1.ListOptions{})
```
Comment (jsafrane) [inline pkg/controller/controller.go]: IMo you should use informer here, we don't want to load the API server with every PVC being cloned.

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -997,6 +1070,150 @@ func (p *csiProvisioner) Delete(volume *v1.PersistentVolume) error {
...
+	// Check for pvc state with DataSource pointing to claim
+	for _, pvc := range pvcList.Items {
+		if pvc.Spec.DataSource == nil {
+			continue
+		}
+
+		if pvc.Spec.DataSource.Kind == pvcKind &&
+			pvc.Spec.DataSource.Name == claim.Name &&
+			pvc.Status.Phase == v1.ClaimPending {
+			return apierrs.NewBadRequest(fmt.Sprintf("PVC '%s' is in pending state, cloning in progress", pvc.Name))
```
Comment (jsafrane) [inline pkg/controller/controller.go]: Why `apierrs.NewBadRequest`? Simple `fmt.Errorf()` would probably do.

File: pkg/controller/controller.go [hunk 5]
```diff
@@ -997,6 +1070,150 @@ func (p *csiProvisioner) Delete(volume *v1.PersistentVolume) error {
...
+		}
+	}
+
+	// Remove clone finalizer
+	// need to get the pvc again because the delete has updated the object with a deletion timestamp
+	claim, err = p.client.CoreV1().PersistentVolumeClaims(claim.Namespace).Get(claim.Name, metav1.GetOptions{})
```
Comment (jsafrane) [inline pkg/controller/controller.go]: I don't understand why extra `Get()` is necessary. `claim` already has `DeletionTimestamp`. Yes, it could have changed since the informer got update event from API server, but it could have changed also since you get it from API server here and use the information few lines below.

File: pkg/controller/controller.go [hunk 6]
```diff
@@ -997,6 +1070,150 @@ func (p *csiProvisioner) Delete(volume *v1.PersistentVolume) error {
...
+
+	finalizers := make([]string, 0)
+	for _, finalizer := range claim.ObjectMeta.Finalizers {
+		if finalizer != pvcCloneFinalizer {
+			finalizers = append(finalizers, finalizer)
+		}
+	}
+
+	// Only update the finalizers if we actually removed something
+	if checkFinalizer(claim, pvcCloneFinalizer) {
```
Comment (jsafrane) [inline pkg/controller/controller.go]: You already checked the finalizer above.

File: pkg/controller/controller.go [hunk 7]
```diff
@@ -311,6 +325,43 @@ func NewCSIProvisioner(client kubernetes.Interface,
 	return provisioner
 }
 
+// NewCSIController creates new controller for additional CSI capabilities
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply pkg/controller/controller.go]: And maybe call it CloningProtectionController, until someone adds another responsibility to it?

File: pkg/controller/clone_controller_test.go [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+			)
+
+			go cloningProtector.Run(1, context.TODO().Done())
+
+			// Simulate Delete behavior
+			claim := tc.cloneSource.(*v1.PersistentVolumeClaim)
+			clientSet.CoreV1().PersistentVolumeClaims(claim.Namespace).Update(pvcDeletionMarked(claim))
+
+			// Wait for couple reconciles for controller to adjust finalizers
+			time.Sleep(2 * time.Second)
```
Comment (jsafrane) [inline pkg/controller/clone_controller_test.go]: Is it necessary to wait for a couple of reconciles? For an unit test we typically call `syncClaim()` and don't run whole controller loop with workqueues.

(Yes, there will be some functions not covered by unit tests.)

Comment (Danil-Grigorev) [reply pkg/controller/clone_controller_test.go]: No, not at all. Will fix this, thanks for suggestion. It will also allow to catch an error, which forces the reconcile to run again.

File: pkg/controller/controller.go (hunk 3 above)
Comment (Danil-Grigorev) [reply pkg/controller/controller.go]: Fixed.

File: pkg/controller/controller.go (hunk 2 above)
Comment (Danil-Grigorev) [reply pkg/controller/controller.go]: Thanks, didn't notice that!

File: pkg/controller/controller.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (Danil-Grigorev) [reply pkg/controller/controller.go]: Sounds good, just added tests which should describe the behaviour and moved it all to a separate file.

File: cmd/csi-provisioner/csi-provisioner.go [hunk 9]
```diff
@@ -234,16 +247,23 @@ func main() {
...
 
+	csiClaimController := ctrl.NewCloningProtectionController(
+		clientset,
+		claimLister,
+		claimInformer,
+		claimQueue,
+	)
+
 	run := func(context.Context) {
-		stopCh := context.Background().Done()
```
Comment (jsafrane) [inline cmd/csi-provisioner/csi-provisioner.go]: Why did you remove `stopCh`?

File: pkg/controller/clone_controller_test.go [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+		},
+		"delete source pvc when destination pvc status is claim pending": {
+			cloneSource: pvcFinalizers(baseClaim(), pvcCloneFinalizer),
+			initialClaims: []runtime.Object{
+				pvcPhase(
+					pvcDataSourceClone(
+						pvcNamed(baseClaim(), dstName),
+						srcName,
+					),
+					v1.ClaimPending)},
```
Comment (jsafrane) [inline pkg/controller/clone_controller_test.go]: It would be more readable if you put `claim` as the last argument - function name + its arguments would be on the same line:

```
pvcPhase(v1.ClaimPending,
    pvcDataSourceClone(srcName,
        pvcNamed(dstName, baseClaim())
```

File: pkg/controller/controller.go [hunk 11]
```diff
@@ -134,6 +134,8 @@ const (
 	annStorageProvisioner = "volume.beta.kubernetes.io/storage-provisioner"
 
 	snapshotNotBound = "snapshot %s not bound"
+
+	pvcCloneFinalizer = "external-provisioner.clone.kubernetes.io/finalizer"
```
Comment (jsafrane) [inline pkg/controller/controller.go]: provisioner.storage.kubernetes.io/cloning-protection

File: cmd/csi-provisioner/csi-provisioner.go (hunk 9 above)
Comment (Danil-Grigorev) [reply cmd/csi-provisioner/csi-provisioner.go]: Accidentally left this change after seeing issues with a similar approach in tests, will undo this.

**Comment by jsafrane** [comment]
Good work and thanks a lot!
/lgtm
/approve

File: pkg/controller/clone_controller.go [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+	defer utilruntime.HandleCrash()
+	defer p.claimQueue.ShutDown()
+
+	claimHandler := cache.ResourceEventHandlerFuncs{
+		AddFunc:    p.enqueueClaimUpadate,
+		UpdateFunc: func(_ interface{}, newObj interface{}) { p.enqueueClaimUpadate(newObj) },
+	}
+	p.claimInformer.AddEventHandlerWithResyncPeriod(claimHandler, controller.DefaultResyncPeriod)
+
+	for i := 0; i < threadiness; i++ {
```
Comment (msau42) [inline pkg/controller/clone_controller.go]: This is going to double the number of goroutines, and default value of worker-threads is 100 (so 200) total. Do we really need that many threads to process this finalizer? @jsafrane

Comment (jsafrane) [reply pkg/controller/clone_controller.go]: hm, maybe not...

Comment (Danil-Grigorev) [reply pkg/controller/clone_controller.go]: I'll address this in a follow up PR. Sorry for the issue. Could 5 be sufficient @msau42?
